Expand sparse.softmax zero nnz tests to cover cases of previously reported FPE.

amjames opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95646

- Test cases with zero `nnz` added for `sparse.log_softmax`.
- Test cases with zero `nnz` for both `sparse.log_softmax` and
`torch.sparse_softmax` expanded to cover the backward pass.

These test additions prove resolution to #95371 and #82107.

Fixes #82107 #95371

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @bhosmer

amjames committed:
Expand sparse.softmax zero nnz tests to cover cases of previously reported FPE.

- Test cases with zero `nnz` added for `sparse.log_softmax`.
- Test cases with zero `nnz` for both `sparse.log_softmax` and
`torch.sparse_softmax` expanded to cover the backward pass.

These test additions prove resolution to #95371 and #82107.

Fixes #82107 #95371

[ghstack-poisoned]

amjames committed:
Update on "Expand sparse.softmax zero nnz tests to cover cases of previously reported FPE."


- Test cases with zero `nnz` added for `sparse.log_softmax`.
- Test cases with zero `nnz` for both `sparse.log_softmax` and
`torch.sparse_softmax` expanded to cover the backward pass.

These test additions prove resolution to #95371 and #82107.

Fixes #82107 #95371

cc alexsamardzic nikitaved pearu cpuhrsch bhosmer

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse.py:
```diff
@@ -3481,6 +3481,25 @@ def test_softmax_zero_nnz(self, device, dtype):
+        t = t.requires_grad_()
+
+        def fn(x):
+            return torch.sparse.softmax(x, 0).to_dense()
```
Can we use straight gradcheck now that https://github.com/pytorch/pytorch/pull/95405 merged?

cpuhrsch replied:
Perhaps even enable OpInfo coverage?

amjames replied:
Gradcheck still does not support sparse outputs, so this wrapper is still needed.

cpuhrsch replied:
@pearu - Can we extend gradcheck to work for this case too?

pearu replied:
This works:
```python
>>> torch.autograd.gradcheck(lambda x: torch.sparse.softmax(x, 0).to_dense(), t, masked=True, check_sparse_nnz=True)
True
```

Here's a list of tasks for extending gradcheck:
1. Eliminate `check_sparse_nnz=` kw argument as redundant. This is an easy fix.
2. Allow op to return sparse output. I tried to disable the check --- there appear serious issues that need to be resolved to get sparse output working.

nikitaved replied:
Also nit: can we enable `float` dtype as well?

pearu replied:
@nikitaved float32 is typically disabled for gradcheck tests because the numerical Jacobian result is often too inaccurate. To make the test work, tweaking of default eps/atol/rtol parameters is required.

nikitaved replied:
These are trivial cases with no computation, so it should work for any dtype I reckon.

amjames replied:
Also curious if the "real" sparse softmax tests need to do all the stuff they are doing with computing and comparing the full Jacobian. Can't we just use gradcheck there as well? If we simplify that test then this just becomes an additional shape + nnz to pass as a test case.

pearu replied:
I suggest keeping "the real sparse softmax tests" because it provides a closed form jacobian (as defined in `softmax_jacobian_analytic`) that would be the thrid independent approach of testing "sparse softmax" gradients together with the numerical and analytical jacobians from gradcheck.

amjames committed:
Update on "Expand sparse.softmax zero nnz tests to cover cases of previously reported FPE."


- Test cases with zero `nnz` added for `sparse.log_softmax`.
- Test cases with zero `nnz` for both `sparse.log_softmax` and
`torch.sparse_softmax` expanded to cover the backward pass.

These test additions prove resolution to #95371 and #82107.

Fixes #82107 #95371

cc alexsamardzic nikitaved pearu cpuhrsch bhosmer

[ghstack-poisoned]

pearu commented on test/test_sparse.py:
```diff
@@ -3503,6 +3503,25 @@ def test_softmax_zero_nnz(self, device, dtype):
+
+        def fn(x):
+            return torch.sparse.softmax(x, 0).to_dense()
+        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True)
```
Use:
```suggestion
        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True, masked=True)
```

pearu commented on test/test_sparse.py:
```diff
@@ -3503,6 +3503,25 @@ def test_softmax_zero_nnz(self, device, dtype):
+
+        def fn(x):
+            return torch.sparse.log_softmax(x, 0).to_dense()
+        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True)
```
Use:
```suggestion
        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True, masked=True)
```

pearu reviewed (changes requested):
`masked=True` is required when autograd is applied on operations from `torch.sparse` namespace.

Otherwise, LGTM!

nikitaved commented on test/test_sparse.py:
```diff
@@ -3503,6 +3503,25 @@ def test_softmax_zero_nnz(self, device, dtype):
+        torch.autograd.gradcheck(fn, (t), check_sparse_nnz=True)
+
+    @dtypes(torch.float)
+    def test_log_softmax_zero_nnz(self, device, dtype):
+        t = torch.sparse_coo_tensor([[]], [], (3,), device=device, dtype=dtype)
```
Are zero sizes implying zero nnz tested as well?

pearu replied:
There seems to be no issue with zero sizes:
```python
>>> t = torch.sparse_coo_tensor([[], []], [], (0, 3), dtype=torch.float64)
>>> t.requires_grad_()
tensor(indices=tensor([], size=(2, 0)),
       values=tensor([], size=(0,)),
       size=(0, 3), nnz=0, dtype=torch.float64, layout=torch.sparse_coo,
       requires_grad=True)
>>> torch.autograd.gradcheck(lambda x: torch.sparse.softmax(x, 0).to_dense(), t, masked=True, check_sparse_nnz=True)
True
```

nikitaved replied:
But is it tested? Looks like a great opportunity to cover degenerate cases.

amjames committed:
Update on "Expand sparse.softmax zero nnz tests to cover cases of previously reported FPE."


- Test cases with zero `nnz` added for `sparse.log_softmax`.
- Test cases with zero `nnz` for both `sparse.log_softmax` and
`torch.sparse_softmax` expanded to cover the backward pass.

These test additions prove resolution to #95371 and #82107.

Fixes #82107 #95371

cc alexsamardzic nikitaved pearu cpuhrsch bhosmer

[ghstack-poisoned]

amjames committed:
Update on "Expand sparse.softmax zero nnz tests to cover cases of previously reported FPE."


- Test cases with zero `nnz` added for `sparse.log_softmax`.
- Test cases with zero `nnz` for both `sparse.log_softmax` and
`torch.sparse_softmax` expanded to cover the backward pass.

These test additions prove resolution to #95371 and #82107.

Fixes #82107 #95371

cc alexsamardzic nikitaved pearu cpuhrsch bhosmer

[ghstack-poisoned]

pearu reviewed (approved):
LGTM! Thanks, @amjames!

nikitaved reviewed (approved):
👍

amjames commented:
@pytorchbot merge
